---
# Visual Studio Code (VSCode)
---

How to use this notebook:

- First, work through the prerequisites in the whole of [1. Prerequisites](#1-prerequisites).
- Then, work though [2. VSCode Basics](#2-vscode-basics).
- Next, work through [3. Jupyter Notebook Basics](#3-jupyter-notebook-basics).
- Finally, work though [4. Google CoLab Basics](#4-google-colab-basics).
- If you want to clean up any files created by this notebook, you can also work through [5. Cleanup](#5-cleanup).

This notebook covers:

- [1. Prerequisites](#1-prerequisites) 
  - [1.1 Software](#11-software)
  - [1.2 VSCode Extensions](#12-vscode-extensions)
  - [1.3 Accounts](#13-accounts)
  - [1.4 Course Repository](#14-course-repository)
  - [1.5 Python (Conda) Environment](#15-python-conda-environment)
  - [1.6 Using Built-in Cell Magic `%%writefile`](#16-using-built-in-cell-magic-writefile)
- [2. VSCode Basics](#2-vscode-basics)  
  - [2.1 User Interface](#21-user-interface)
  - [2.2 VSCode Debug Configuration](#22-vscode-debug-configuration)
  - [2.3 Running a Python Program in a Notebook Code Cell](#23-running-a-python-program-in-a-notebook-code-cell)
  - [2.4 Debugging a Single-file Python Program](#24-debugging-a-single-file-python-program)
  - [2.5 Debugging a Multi-file Python Program](#25-debugging-a-multi-file-python-program)
- [3. Jupyter Notebook Basics](#3-jupyter-notebook-basics)
  - [3.1 Creating a Jupyter Notebook](#31-creating-a-jupyter-notebook)
  - [3.2 Choosing the Python Interpreter (Environment)](#32-choosing-the-python-interpreter-environment)
  - [3.3 Creating a Markdown Cell](#33-creating-a-markdown-cell)
  - [3.4 Creating a Code Cell](#34-creating-a-code-cell)
- [4. Google CoLab Basics](#4-google-colab-basics)
  - [4.1 Opening a Jupyter Notebook in Google CoLab](#41-opening-a-jupyter-notebook-in-google-coLab)
- [5. Cleanup](#5-cleanup)

---
# 1. Prerequisites
---

## 1.1 Software

In this course, the following software needs to be installed on your computer:

- [Visual Studio Code](https://code.visualstudio.com)
- [Git](https://git-scm.com/downloads)
- [Miniconda](https://docs.anaconda.com/miniconda/install/#quick-command-line-install)
- [Python](https://www.python.org) (optional) - comes pre-installed on Linux and Mac

To verify all necessary software has been installed, run the cell below by placing the cursor in the cell and pressing `Shift + Enter`.
- You should see the name of each tool, together with a version number.
- Note that Python isn't required, since we will be using Conda environments with Python installed inside the environment.

In [19]:
version = !code --version
print(f"vscode {version[0]}")
!git --version
!conda --version
!python --version

vscode 1.103.0
git version 2.43.0
conda 25.3.1
Python 3.12.11


If you don't see the print-out of a version for a specific tool, make sure the path to your tool's folder has been added to your [`PATH` environment variable](https://gist.github.com/nex3/c395b2f8fd4b02068be37c961301caa7).

---
## 1.2 VSCode Extensions

You also need to install the necessary VSCode extensions (plugins).
- VSCode extensions add additional functionality to VSCode.
- The two extensions we need in this course are:
  - [Python](https://marketplace.visualstudio.com/items?itemName=ms-python.python) for working with Python in VSCode.
  - [Jupyter](https://marketplace.visualstudio.com/items?itemName=ms-toolsai.jupyter) for working with Jupyter Notebooks in VSCode.

Run the cell below to install the extensions (make sure the cell is selected, then press `Shift + Enter`).

In [20]:
!code --install-extension ms-toolsai.jupyter --force
!code --install-extension ms-python.python --force

Installing extensions...
Extension 'ms-toolsai.jupyter' is already installed.
Installing extensions...
Extension 'ms-python.python' is already installed.


---
## 1.3 Accounts

You will also need a GitHub account to download content for this course and to share code with your group members.

- If you don't already have an account, visit the [GitHub](https://github.com) web site and [Sign up](https://github.com/signup) for an account.

Also make sure your have initialized Git with your email address, your name, and have set the default branch to main.
- Run the cell below.
- If any of the settings described above are missing:
  - Comment out the lines below.
  - Enter your own email address and name.
  - Run the cell again.

In [25]:
!git config --global --list
# !git config --global user.email "your.email@example.com"
# !git config --global user.name "Your Name"
# !git config --global init.defaultBranch main

user.email=patrick.gabrielsson@hb.se
user.name=Patrick Gabrielsson
credential.helper=store
init.defaultbranch=main


---
## 1.4 Course Repository

The GitHub repository [C1VI1B_2025](https://github.com/paga-hb/C1VI1B_2025) for this course can be cloned to your computer by running the cell below:
- It will clone the repository to the subfolder `dataviz` in the current folder.
- The repository contains all the workshops, the assignment, and the project description for this course.

In [ ]:
!git clone https://github.com/paga-hb/C1VI1B_2025.git dataviz
!cd dataviz

Cloning into 'dataviz2'...
remote: Enumerating objects: 63, done.
remote: Counting objects: 100% (63/63), done.
remote: Compressing objects: 100% (57/57), done.
remote: Total 63 (delta 31), reused 13 (delta 1), pack-reused 0 (from 0)
Receiving objects: 100% (63/63), 20.72 KiB | 964.00 KiB/s, done.
Resolving deltas: 100% (31/31), done.


---
## 1.5 Python (Conda) Environment

When we create a new Python project, we also want to create a new Python environment for it. In this course, we will use Miniconda to create our Python environments.
- The command `conda create -y -p ./.conda python=3.12`
  - Creates a folder called `.conda` in the current folder.
  - Then it installs Python version 3.12 in it (including the `pip` package manager).
- Once the Python environment has been created, the command `conda activate ./.conda`
  - Activates the Python environment in the terminal.
    - Your terminal should give you a hint that you are using the Python interpreter inside that environment.
  - This means all Python commands are mapped to the Python interpreter inside the environment.
- In a new Python environment, we want to run the command `python -m pip install --upgrade pip`
  - To update the `pip` package manager (inside the Python environment) to the latest version.
- Now we can install any Python packages needed for our Python project with the command `pip install <package>`
  - Where `<package>` is the name of the Python package.
  - You can also install multiple packages by separating them with a space, e.g. `pip install <package1> <package2>`

For the workshops, the assignment, and the project in this course, we need to install the following packages in our Python environment.
- `ipykernel`, `jupyter`, and `pylance` for running Jupter Notebooks, and for Python Intellisense.
- `numpy`, and `pandas` for working with data in Python.
- `matplotlib`, `seaborn`, `bokeh`, and `plotly` for data visualization.
- `dash`, `dash-bootstrap-components`, `openpyxl`, `lxml`, and `pycountry` for interactive data visualization apps.

In summary, for Python projects in this course, we want to use the Python environment below:

```bash
conda create -y -p ./.conda python=3.12
conda activate ./.conda
python -m pip install --upgrade pip
pip install ipykernel jupyter pylance numpy pandas matplotlib seaborn bokeh plotly
pip install dash dash-bootstrap-components openpyxl lxml pycountry
```

Lastly, when running Jupyter Notebooks in this course, we also want to select this Python environment inside the Notebook.
- This is done by clicking the `Select Interpreter` button in the top right of a Jupyter Notebook.

---
## 1.6 Using Built-in Cell Magic `%%writefile`

- The cell magic `%%writefile filename`, writes the contents of a notebook cell to the specified `filename` (or file path).
  - This functionality is built-in to Jupyter Notebooks (it's not an extension).
  - We can use it to write any code cell contents to a file in the file system.
  - Let's write some Python code to the file `main.py`.
- Run the cell below.
- Then inspect the resulting file `main.py`, which contains the code cell's contents (except for the cell magic command `%%writefile filename`).

In [4]:
%%writefile main.py
def main():
    print("Hello world!\n")
    return 0

if __name__ == "__main__":
    main()

Overwriting main.py


We can run the Python program from a Notebook cell by invoking the Python Interpreter and passing in the filename as an argument.
- When an exclamation mark `!` is used in a code cell, anything that follows it is run in the underlying operating system's shell (terminal), where any output is show in the cell's output.

Run the cell below to execute the Python program in main.py, and observe the program output.

In [6]:
!python main.py

Hello world!



---
# 2. VSCode Basics
---

## 2.1 User Interface

Microsoft's [Getting Started Tutorial](https://code.visualstudio.com/docs/getstarted/getting-started) is great if you haven't used Visual Studio Code before, but we will explore Visual Studio Code together in this notebook.

Visual Studio Code's [user interface](https://code.visualstudio.com/docs/getstarted/userinterface) is quite elaborate, but let's explore its most important parts below.

<style>
    .container {
        width: 98%;
        margin-left: 0; /* Push the container to the left */
        margin-right: auto; 
    }
    .text-image {
        margin-bottom: 35px; /* Space between sections */
        overflow: hidden; /* Ensure image stays within the container */
    }
    .text {
        text-align: justify; /* Justify the text for better readability */
    }
    .image {
        float: right; /* Float the image to the right */
        margin-left: 25px; /* Space between image and text */
        margin-bottom: 10px; /* Space between image and text */
        max-width: 50%; /* Limit image size */
        height: auto; /* Maintain aspect ratio */
    }
</style>

<div class="container">
    <div class="text-image"> <!-- Explorer View -->
        <img class="image" src="../images/explorer-view.png">
        <div class="text">
            <h4>General Layout and the (File) <a href="https://code.visualstudio.com/docs/getstarted/userinterface">Explorer View</a></h4>
            <p>
                The <em>Main Menu</em> is located along the top of VSCode's UI, where a <a href="https://code.visualstudio.com/docs/editor/workspaces">Workspace</a> folder can be opened with <b>File -> Open Folder</b>. The currently open workspace folder's name is displayed in the <em>Explorer</em>. When you start working on a development project, you should <b>always open the folder that contains your project via the main menu, or by moving to the folder in your terminal and entering the command <code>code .</code></b> (the word <em>code</em>, followed by a space, and then a <em>dot</em>).
            </p>
            <p>
                The first menu choice <img src="../images/explorer-view-icon.png"/> in VSCode's <em>Activity Bar</em> is the <em>Explorer</em>, which displays all files in your current workspace in the <em>Primary Side Bar</em>.
            </p>
            <p>
                When you double-click a file in the <em>Primary Side Bar</em>, it's contents are displayed in a tab under the <em>Editor Groups</em>. The <em>Editor Groups</em> can display one <em>Editor Group</em> or multiple <em>Editor Groups</em> side-by-side. Each <em>Editor Group</em> can have one or more files open, each in a separate tab. A new <em>Editor Group</em> can be created by using the icons in the top right of VSCode's UI (which is useful for e.g. viewing two files side-by-side at the same time).
            </p>
            <p>
                Below the <em>Editor Groups</em>, the <em>Panel</em> is displayed. The <em>Panel</em> has multiple tabs, e.g. the tabs <em>Problems</em>, <em>Output</em>, <em>Terminal</em> and <em>Ports</em>. Currently, the <em>Terminal</em> tab is selected, which displays VSCode's built-in terminal. Additional terminals can be created by clicking the <b>+</b> icon in the top-right of the <em>Panel</em>. The <em>Output</em> tab displays output genereated by an executing program, or the compiler. The <em>Panel</em> can be toggled between hidden and displayed by using <b>Ctrl + J</b> (or <b>Cmd + J</b> on a Mac).
            </p>
            <p>
                The <em>Status Bar</em> is located along the bottom of VSCode's UI, and displays various status information, e.g. the carret's current line and column position in the currently active opened file.
            </p>
        </div>
    </div>
</div>
<div class="container">
    <div class="text-image"> <!-- Source Control View -->
        <img class="image" src="../images/source-control-view.png" />
        <div class="text">
            <h4>The Search View and the <a href="https://code.visualstudio.com/docs/sourcecontrol/overview">Source Control View</a></h4>
            <p>
                The second menu choice <img src="../images/search-view-icon.png"/> in VSCode's <em>Activity Bar</em> displays the <em>Search</em> view, which can be used to search for strings among the workspace's files.
            </p>
            <p>
                The third menu choice <img src="../images/source-control-icon.png"/> in VSCode's <em>Activity Bar</em> displays the <em>Source Control</em> view, which can be used to stage and commit changed files to GitHub. A modified file will have an <b>M</b> to the right of it's file name. In general, you would add modified files to the <b>Staged Changes</b>, fill in a comment at the top of the view, and then click the <b>Commit</b> button. The <em>Source Control</em> view can also be used to compare files, and to resolve merge conflicts.
            </p>
        </div>
    </div>
</div>
<div class="container">
    <div class="text-image"> <!-- Run and Debug View -->
        <img class="image" src="../images/run-and-debug-view.png" />
        <div class="text">
            <h4>The <a href="https://code.visualstudio.com/docs/editor/debugging">Run and Debug View</a></h4>
            <p>
                The <em>Run and Debug Icon</em> <img src="../images/run-and-debug-view-icon.png"/> opens the <em>Run and Debug</em> view, which can be used to run and debug code in VSCode. Various debug profiles can be defined, but in genereal you would press <b>F5</b> to start a debug session, choose <b>Run -> Start Debugging</b> from VSCode's main menu, or click the green arrow in the top-left of the <em>Run and Debug</em> view. 
            </p>
            <p>
                You can toggle <em>break points</em> on and off in your code by pressing <b>F9</b> on the currently selected row, or by clicking in the editor's left margin, which will toggle a red circle on or off.
            </p>
            <p>
                When you start a debug session and hit a <em>break point</em>, the <em>Debug Side Bar</em> will display the <b>VARIABLES</b> in scope and their associated values. There is also a <b>WATCH</b> section where you can define watches for various variables. You can also view the <b>CALL STACK</b> and the <b>BREAKPOINTS</b> set in your code.
            </p>
            <p>
                At the top of the editor, you will find a number of buttons you can use to debug your code, which are (from left to right), <b>Continue</b> (to continue execution until the next breakpoint is hit), <b>Pause</b> (to pause execution), <b>Step Over</b> (to step over a function/method call without entering its body), <b>Step In</b> (to step into a function's/method's body), <b>Step Out</b> (to step out of a function's/method's body), <b>Restart</b> (to restart the debug session), and <b>Stop</b> (to stop the debug session).
            </p>
            <p>
                At the bottom of the editor, you will find the <em>Debug Console Panel</em> that will display various messages from the compiler.
            </p>
        </div>
    </div>
</div>
<div class="container">
    <div class="text-image"> <!-- Extensions Marketplace View -->
        <img class="image" src="../images/extensions-view.png" />
        <div class="text">
            <h4>The <a href="https://code.visualstudio.com/docs/editor/extension-marketplace">Extensions Marketplace View</a></h4>
            <p>
                The <em>Extensions Icon</em> <img src="../images/extensions-view-icon.png"/> opens the <em>Extensions Marketplace</em> view, which shows the currently installed <em>Visual Studio Code Extensions</em>. It can also be used to search for extensions from the <em>Extension Marketplace</em>, by entering a serch term into the search bar at the very top of the <em>Extensions Side Bar</em>. 
            </p>
            <p>
                When you select an extension in the <em>Extensions Side Bar</em>, information about the extension is displayed to the right in VSCode's UI. In the right margin, under <em>More Info</em>, you will find the <b>Identifier</b> for the extension. If you want to install the extension using the terminal, you can use the command <code>code --install-extension [identifier]</code>, where <code> [identifier]</code> is the <b>Identifier</b> displayed under <em>More Info</em>. To install the selected extension via the UI, click the <b>Install</b> button next to the extension (to uninstall it, click the <b>Uninstall</b> button). 
            </p>
            <p>
                To develop Python applications in this course, you need the following extensions (as a bare minimum):
                <p>
                    <code>code --install-extension ms-python.python</code><br/>
                    <code>code --install-extension ms-toolsai.jupyter</code><br/>
                </p>
            </p>
        </div>
    </div>
</div>
<div class="container">
    <div class="text-image"> <!-- Test Explorer View -->
        <img class="image" src="../images/test-explorer-view.png" />
        <div class="text">
            <h4>The <a href="https://code.visualstudio.com/docs/editor/testing">Test Explorer View</a></h4>
            <p>
                The <em>Test Explorer Icon</em> <img src="../images/test-explorer-view-icon.png"/> opens the <em>Test Explorer</em> view, which can can be used run and debug tests, if you have any test projects in your current workspace. 
            </p>
            <p>
                The <b>TEST EXPLORER</b> <em>Side Bar</em> displays the tests from all test projects in your current workspace, where you can choose to run or debug individual tests, or all tests in your current workspace. A red icon next to a test indicates the test has failed, and a green icon indicates that a test has passed. The <b>TEST COVERAGE</b> section displays how much of your code is covered by your tests.
            </p>
            <p>
                The test results are also displayed in the <b>TEST RESULTS</b> panel below the editor.
            </p>
            <p>
                When debugging a test, the test execution will pause execution when a <em>break point</em> is hit, which gives you the opportunity to display the <em>expected value</em> and the <em>actual value</em> of your test code, side-by-side.
            </p>
            <p>
                You can also start a test from your code in the editor, by clicking the arrow displayed in the left margin next to your test code.
            </p>
        </div>
    </div>
</div>

In [2]:
import os
os.makedirs(".vscode", exist_ok=True)

---
## 2.2 VSCode Debug Configuration

- To develop Python programs in VSCode, we need to configure one VSCode workspace configuration file.
  - In the file `launch.json` we can configure various debug options, such as debugging single-file or multi-file Python programs.
- VSCode workspace configuration files (`.json`) are stored in the subfolder `.vscode`.

The easiest way to create a `launch.json` file for a new Python project, is to:

- Switch to the **Run and Debug** view (Windows/Linux: `Ctrl + Shift + D`, MacOS: `Cmd + Shift + D`).
- In the Activity Bar for the **Run and Debug** view, click the link `create a launch.json file`.
- Choose **Select Debugger** to `Python Debugger`.
- Choose **Select a debug configuration** to `Python File Debug the currectly active Python file`.
- The file `launch.json` will be created in subfolder `.vscode`.
- Save the file `launch.json`.

The file contains the following settings:
- `"name": "Python Debugger: Current File"` which is the name of this debug configuration.
- `"type": "debugpy"` which configures this configuration to run the Python debugger.
- `"request": "launch"` which launches the Python debugger when this debug configuration is run.
- `"program": "${file}"` which is the currently open file in the editor (`${file}` is replaced with the actual filename).
- `"console": "integratedTerminal"` which redirects program input/output to VSCode's integrated terminal.

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        
        {
            "name": "Python Debugger: Current File",
            "type": "debugpy",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal"
        }
    ]
}
```

This debug configuration can be used to debug a single-file Python program (the `.py` file currently open in the editor). Let's also add a configuration that let's us debug a multi-file Python program.

- Open the file `launch.json` in subfolder `.vscode`, and make the following changes (then save the file).

```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        
        {
            "name": "Python Debugger: Current File",
            "type": "debugpy",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal"
        },
        {
            "name": "Python Debugger: main.py",
            "type": "debugpy",
            "request": "launch",
            "program": "${workspaceFolder}/main.py",
            "console": "integratedTerminal"
        }
    ]
}
```

Notice the new debug configuration has the following settings:
- `"name": "Python Debugger: main.py"` which is the name of this debug configuration.
- `"type": "debugpy"` which configures this configuration to run the Python debugger.
- `"request": "launch"` which launches the Python debugger when this debug configuration is run.
- `"program": "${workspaceFolder}/main.py"` which is the file `main.py` in the root workspace folder.
- `"console": "integratedTerminal"` which redirects program input/output to VSCode's integrated terminal.

This debug configuration can be used to debug a multi-file Python program with the programs entry point in the file `main.py` in the root workspace folder.

The current debug configuration (single-file or multi-file) can be chosen in the **Run and Debug** view (choose the debug configuration's name in the dropdown listbox, the press the green *Run* button or press `F5`).

---
## 2.3 Running a Python Program in a Notebook Code Cell

Jupyter Notebooks natively support Python, so we can write any Python program spanning the code cells in a Jupyter Notebook.

Run the cell below to execute the Python program that it contains.

In [1]:
def main():
    print("Hello World!\n")

if __name__ == "__main__":
    main()

Hello World!



---
## 2.4 Debugging a Single-file Python Program

Let's see `launch.json` in action for a single-file (`.py`) Python program.

- First, let's create the file `main.py` in the cell below.

In [2]:
%%writefile main.py
import math

def main():
    radius = 1.0;
    print(f"circleArea = {math.pi} * {radius} * {radius} = {circleArea(radius)}\n")
    return 0

def circleArea(radius):
    return math.pi * radius * radius;

if __name__ == "__main__":
    main()

Overwriting main.py


- Now, let's debug the file `main.py`.
  - Open the file `main.py` in VSCode's editor.
  - Set a breakpoint on the two `return` statements (`F9`).
  - Switch to the `Run and Debug` view (Linux/Windows: `Ctrl + Shift + D`, Mac: `Cmd + Shift + D`).
  - In the drop-down combobox, select the launch configuration `Python Debugger: Current File`.
  - Click the green `Play` icon.
  - Use the debug toolbar in the top-middle of VSCode to debug the code.
    - Notice the debugger stops at the breakpoints.
    - Notice you can view local variables, watch variables, view the call stack, and toggle breakpoints in the `Run and Debug` view.
  - Stop debugging (red `Square` icon in the debug toolbar).
- Next, look at the status bar (at the bottom of VSCode) where you will see the name of the launch configuration `Python Debugger: Current File`.
  - Click on it, and select `Python Debugger: Current File` again (make sure `main.py` is the active file in the editor, not the notebook).
    - This is an alternative method to start a debug session.
  - Stop debugging.
- Press `F5` (make sure `main.py` is active in VSCode's editor), which is a third alternative to launch the debugger.
  - Stop debugging.

- Remeber, you can always run a `.py` file in your filesystem from a Notebook cell using the command below.

In [4]:
!python main.py

circleArea = 3.141592653589793 * 1.0 * 1.0 = 3.141592653589793



**Note**

- You can select the Python interpreter (i.e. Python Environment) to use for running a `.py` file, by opening any `.py` file and clicking the *Interpreter Picker* in the bottom right of VSCode (in the Status Bar).
- You can create files and folders in the **Explorer** view (Windows/Linux: `Ctrl + Shift + E`, MacOS: `Cmd + Shift + E`) by hovering the mouse over the Activity Bar and clicking the buttons at the top of the Activity Bar.

---
## 2.5 Debugging a Multi-file Python Program

- Let's see `launch.json` in action for a multi-file (`*.py`) Python program.
  - First we will create a module `trigonometry.py` in an `arithmetic` folder, which will be our package.
    - We will place the code for `circleArea` in this module.
  - Then we will import `circleArea` into `main.py` and produce the same output as before.

- Run the four cells below to create:
  - The folder `arithmetic` in the root workspace folder (this is our Python package).
  - The empty file `__init__.py` in the folder `arithmetic` (this is what makes `arithmetic` a Python package).
  - The file `trigonometry.py` in the folder `arithmetic` (this is our module within our package).
  - The file `main.py` in the root workspace folder (this file contains our main Python program).

In [5]:
import os
os.makedirs("../arithmetic", exist_ok=True)

In [8]:
%%writefile ../arithmetic/__init__.py
pass

Writing ../arithmetic/__init__.py


In [12]:
%%writefile ../arithmetic/trigonometry.py
import math

def circleArea(radius):
    return math.pi * radius * radius

Overwriting ../arithmetic/trigonometry.py


In [13]:
%%writefile ../main.py
import math
from arithmetic.trigonometry import circleArea

def main():
    radius = 1.0;
    print(f"circleArea = {math.pi} * {radius} * {radius} = {circleArea(radius)}\n")
    return 0

if __name__ == "__main__":
    main()

Overwriting ../main.py


- Debug the multi-file Python program.
  - Open `main.py` and `trigonometry.py` in the `arithmetic` folder and set breakpoints on the two `return` statements.
  - Switch the the `Run and Debug` view, select `Python Debugger: main.py` from the drop-down list:
    - Then click the green `Play` icon.
    - Or, press `F5`.
  - You can also select `Python Debugger: main.py` from the status bar (at the bottom of VSCode).
  - Stop debugging.

- Remeber, you can always run a multi-file Python program from a Notebook cell with the command below.

In [14]:
!python ../main.py

circleArea = 3.141592653589793 * 1.0 * 1.0 = 3.141592653589793



**Note**

- You can select the Python interpreter (i.e. Python Environment) to use for running a `.py` file, by opening any `.py` file and clicking the *Interpreter Picker* in the bottom right of VSCode (in the Status Bar).

---
# 3. Jupyter Notebook Basics
---

## 3.1 Creating a Jupyter Notebook

A Jupyter Notebook is any file with the file extension `.ipynb`.

Let's create the Jupyter Notebook `main.ipynb` by running the cell below.

In [1]:
open("main.ipynb", "w").close()

Now, right-click the file in VSCode's Explorer (Windows/Linux: `Ctrl + Shift + E`, MacOS: `Cmd + Shift + E`) and choose `Open to the Side`.

---
## 3.2 Choosing the Python Interpreter (Environment)

The first thing you do when opening a Jupyter Notebook, is to choose a Python Interpreter (Environment).
- Right-click `Select Kernel` in the top right of the Notebook.
- Choose the Python Interpreter in the `.conda` folder.
- Now the text `Select Kernel` should have changed to something like `.conda (Python 3.12)`.

---
## 3.3 Creating a Markdown Cell

To learn about Markdown, visit [https://www.markdownguide.org](https://www.markdownguide.org), but let's cover some basics below.

In the Jupyter Notebook `main.ipynb`:

- Click the menu option `+ Markdown` to add a new Markdown Cell to the Notebook.

- In the Markdown Cell, add the text below, then press `shift + enter` to execute the cell:

```markdown
# Heading One
## Heading Two
### Heading Three

This is normal text.

**This text is in bold.**

*This text is in italic.*

~~This text is in strikethrough~~

`This is inline code`

```This is a block of code that can span multiple lines```

Here's an unordered list:

- First listitem.
- Second listitem.
  - Nested listitem.
- Third listitem.

Here's an ordered list:

1. First listitem.
2. Second listitem.
   a) Nested listitem.
3. Third listitem.

[This is a link to https://www.google.com](https://www.google.com)

This is a link to a picture:

![Cute kitten](../images/explorer-view-icon.png)

This is a table:

| Name     | Age | City       |
|----------|-----|------------|
| Alice    | 25  | New York   |
| Bob      | 30  | London     |
| Charlie  | 22  | San Diego  |

This is a horizontal line:

---

> This is a blockquote.  
> It can span multiple lines.  
>> You can even nest blockquotes with double `>>`.

This is a task list:

- [x] Write the draft
- [ ] Review notes
- [ ] Submit final version
```

---
## 3.4 Creating a Code Cell

A Code Cell contains Python code.

In the Jupyter Notebook `main.ipynb`:

- Click the menu option `+ Code` to add a new Code Cell to the Notebook.

- In the Code Cell, add the text below, then press `shift + enter` to execute the cell:

  ```python
  print("Hello World!")
  ```
- The cell output should show the text `Hello World!`

To learn more about Jupyter Notebooks, visit [https://docs.jupyter.org](https://docs.jupyter.org)

---
# 4. Google CoLab Basics
---

## 4.1 Opening a Jupyter Notebook in Google CoLab

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/paga-hb/C1VI1B_2025/blob/main/workshop1/vscode.ipynb)

To open a Jupyter Notebook in Google CoLab:

- Visit [https://colab.research.google.com/](https://colab.research.google.com).
- Choose to `Upload` a Notebook (`File -> Upload notebook`).
- Select the Notebook to upload.
- Then choose `File -> Save a Copy in Drive`.
- You will recognize the menu choices for Markdown Cells and Code Cells.
  - Most Python packages are pre-installed in Google CoLab.
  - You can always istall more Python Packages in a code cell with `!pip install <package>`
- When you are done with your Notebook, you can `File -> Download -> Download .ipynb`

**Note**

You can also open this Notebook in Google CoLab by:
- Clicking the `Open in Colab` badge above.
- Then choose `File -> Save a Copy in Drive`.

To learn more about Google CoLab, visit the [Getting Started](https://colab.research.google.com/#scrollTo=GJBs_flRovLc) guide.

---
# 5. Cleanup
---

Run the cell below to remove any files or folders created by this notebook. 

In [1]:
import os, shutil

dirs = ["../arithmetic", "../.vscode"]
files = ["../main.py", "main.py", "main.ipynb"]

for d in dirs:
    if os.path.exists(d):
        shutil.rmtree(d)

for f in files:
    if os.path.exists(f):
        os.remove(f)